Remote ept point clouds (over HTTP)

NEDJIMAbelgacem opened the pull request:
## Description

This adds the ability to view EPT point cloud from a remote HTTP server like [https://usgs.entwine.io/](url).
The UI is made in a simular way to raster and other layers. The user specifies whether the dataset is a local file or will be loaded over a protocol:
![file_point_cloud](https://user-images.githubusercontent.com/29183781/112172004-9dc2c300-8bf4-11eb-91ad-f28a8186e55f.png)
![remote_point_cloud](https://user-images.githubusercontent.com/29183781/112172019-a0bdb380-8bf4-11eb-83fd-88839331d784.png)

NEDJIMAbelgacem committed:
initial implementation of remote ept

NEDJIMAbelgacem committed:
UI initial implementation for remote ept files

NEDJIMAbelgacem committed:
fix UI

NEDJIMAbelgacem committed:
fix docs

NEDJIMAbelgacem committed:
add license

NEDJIMAbelgacem committed:
fix spelling

NEDJIMAbelgacem committed:
fix cpp check

NEDJIMAbelgacem committed:
add missing ifdef

PeterPetrik commented on src/app/qgisapp.h:
```diff
@@ -1161,7 +1161,7 @@ class APP_EXPORT QgisApp : public QMainWindow, private Ui::MainWindow
      * \since QGIS 3.18
      */
-    QgsPointCloudLayer *addPointCloudLayer( const QString &url, const QString &baseName, const QString &providerKey );
+    QgsPointCloudLayer *addPointCloudLayer( const QString &url, const QString &baseName, const QString &dataSourceType, const QString &providerKey );
```
fix docs with new arg

PeterPetrik commented on src/core/network/qgsfiledownloader.h:
```diff
@@ -108,7 +108,7 @@ class CORE_EXPORT QgsFileDownloader : public QObject
 #endif
 
-  protected:
+  public:
```
is this required?

NEDJIMAbelgacem replied:
I do think it is required, making the desctructor protected in this case will make it impossible to allocate QgsFileDownloader as a local variable and also prevent it from being used in unique_ptr. in the other places where QgsFileDownloader is used, there appears to be a memory leak also. I think we should change it to public.

PeterPetrik commented on src/core/pointcloud/qgseptdecoder.cpp:
```diff
@@ -466,4 +477,27 @@ QgsPointCloudBlock *QgsEptDecoder::decompressLaz( const QString &filename,
          );
 }
 
+#include <QDebug>
```
include to top of the file

PeterPetrik commented on src/core/pointcloud/qgseptdecoder.cpp:
```diff
@@ -466,4 +477,27 @@ QgsPointCloudBlock *QgsEptDecoder::decompressLaz( const QString &filename,
+    const QgsPointCloudAttributeCollection &attributes,
+    const QgsPointCloudAttributeCollection &requestedAttributes )
+{
+  QString filename = "/tmp/node.txt";
```
?

NEDJIMAbelgacem replied:
I just changed this to use QDir::tempPath() + QDir::separator(). In this part I'm writing the downloaded data to a file and then that file content is decompressed. Any suggession how to make this better ?

PeterPetrik replied:
QDir::tempPath() is ok

PeterPetrik commented on src/core/pointcloud/qgseptpointcloudindex.h:
```diff
@@ -47,15 +47,23 @@ class CORE_EXPORT QgsEptPointCloudIndex: public QgsPointCloudIndex
     void load( const QString &fileName ) override;
 
     QgsPointCloudBlock *nodeData( const IndexedPointCloudNode &n, const QgsPointCloudRequest &request ) override;
+    QgsPointCloudBlockHandle *asyncNodeData( const IndexedPointCloudNode &n, const QgsPointCloudRequest &request ) override
```
implementation to cpp. or maybe just in parent class?

PeterPetrik commented on src/core/pointcloud/qgspointcloudblockhandle.h:
```diff
@@ -0,0 +1,59 @@
+ * \ingroup core
+ * \brief Base class for handling loading QgsPointCloudBlock asynchronously
+ *
+ * \note The API is considered EXPERIMENTAL and can be changed without a notice
```
or even private header, does it need to be in public api?

NEDJIMAbelgacem replied:
In case someone wants to use the remote EPT, this needs to be public I think

PeterPetrik commented on src/core/pointcloud/qgspointcloudindex.h:
```diff
@@ -149,6 +152,13 @@ class CORE_EXPORT QgsPointCloudIndex: public QObject
+    //! The access type of the data, local is for local files and remote for remote files (over HTTP)
+    enum AccessType
+    {
+      Local,
```
docs

PeterPetrik commented on src/core/pointcloud/qgspointcloudindex.h:
```diff
@@ -149,6 +152,13 @@ class CORE_EXPORT QgsPointCloudIndex: public QObject
+    enum AccessType
+    {
+      Local,
+      Remote
```
docs

PeterPetrik commented on src/core/pointcloud/qgspointcloudlayer.h:
```diff
@@ -90,6 +90,7 @@ class CORE_EXPORT QgsPointCloudLayer : public QgsMapLayer
      */
     explicit QgsPointCloudLayer( const QString &path = QString(),
                                  const QString &baseName = QString(),
+                                 const QString &dataSourceType = QString(),
```
default "local" ?

PeterPetrik commented on src/core/pointcloud/qgspointcloudlayerrenderer.cpp:
```diff
@@ -177,36 +179,132 @@ bool QgsPointCloudLayerRenderer::render()
+  }
+  else
+  {
+    QElapsedTimer downloadTimer;
+    downloadTimer.start();
+
+    // Async loading of nodes
+    QVector<QgsPointCloudBlock *> blocks( nodes.size(), nullptr );
+    QVector<QgsPointCloudBlockHandle *> blockHandles( nodes.size(), nullptr );
+    QVector<bool> finishedLoadingBlock( nodes.size(), false );
+    QTimer timer;
```
move else block to separate private function

PeterPetrik commented on src/core/pointcloud/qgspointcloudlayerrenderer.cpp:
```diff
@@ -177,36 +179,132 @@ bool QgsPointCloudLayerRenderer::render()
+  if ( pc->accessType() == QgsPointCloudIndex::AccessType::Local )
   {
-    if ( context.renderContext().renderingStopped() )
+    for ( const IndexedPointCloudNode &n : nodes )
     {
-      QgsDebugMsgLevel( "canceled", 2 );
-      canceled = true;
-      break;
+      if ( context.renderContext().renderingStopped() )
+      {
+        QgsDebugMsgLevel( "canceled", 2 );
+        canceled = true;
```
move if block to separate function?

PeterPetrik commented on src/core/pointcloud/qgspointcloudlayerrenderer.cpp:
```diff
@@ -177,36 +179,132 @@ bool QgsPointCloudLayerRenderer::render()
     }
   }
 
+#ifdef QGISDEBUG
+  qDebug() << QStringLiteral( "totals: %1 nodes | %2 points | %3ms" ).arg( nodesDrawn )
+           .arg( context.pointsRendered() )
+           .arg( t.elapsed() );
   QgsDebugMsgLevel( QStringLiteral( "totals: %1 nodes | %2 points | %3ms" ).arg( nodesDrawn )
                     .arg( context.pointsRendered() )
                     .arg( t.elapsed() ), 2 );
+#endif
```
isn't it the same ?

PeterPetrik commented on src/core/pointcloud/qgsremoteeptpointcloudindex.cpp:
```diff
@@ -0,0 +1,521 @@
+
+///@cond PRIVATE
+
+#define PROVIDER_KEY QStringLiteral( "ept" )
+#define PROVIDER_DESCRIPTION QStringLiteral( "EPT point cloud provider" )
```
needed?

PeterPetrik commented on src/core/qgsproviderregistry.h:
```diff
@@ -178,6 +179,12 @@ class CORE_EXPORT QgsProviderRegistry
 
+    /**
+     * Creates new instance of EPT point cloud data provider
+     * \since QGIS 3.10
```
3.20?

PeterPetrik replied:
elso what if build WITH_EPT=False

PeterPetrik commented on src/core/qgsprovidermetadata.h:
```diff
@@ -385,6 +386,12 @@ class CORE_EXPORT QgsProviderMetadata : public QObject
+     * Creates a new instance of EPT point cloud data provider
+     * \since QGIS 3.20
+     */
+    SIP_SKIP virtual QgsEptProvider *createEptDataProvider( const QString &uri, const QString &dataSourceType, const QgsDataProvider::ProviderOptions &options, QgsDataProvider::ReadFlags flags = QgsDataProvider::ReadFlags() ) SIP_FACTORY;
```
what if QGIS is build with WITH_EPT=False?

roya0045 replied:
@PeterPetrik excellent point! Some extra features that are not too common are getting entrenched and complicate the build process, thanks for raising this.

PeterPetrik commented on src/core/pointcloud/qgsremoteeptpointcloudindex.cpp:
```diff
@@ -0,0 +1,521 @@
+
+bool QgsRemoteEptPointCloudIndex::loadSchema( const QByteArray &data )
+{
+  QJsonParseError err;
```
looks like there is quite a lot of duplicate code when there is schema loded async and in sync fashion. wondering if it is possible to reuse the code somehow?

PeterPetrik reviewed (commented):
nice stuff!

nyalldawson commented on python/core/auto_generated/pointcloud/qgspointcloudlayer.sip.in:
```diff
@@ -50,6 +50,7 @@ Constructor for LayerOptions with optional ``transformContext``.
 
     explicit QgsPointCloudLayer( const QString &path = QString(),
                                  const QString &baseName = QString(),
+                                 const QString &dataSourceType = QString(),
```
This is an api break. But can't we avoid the whole "dataSourceType" argument by checking if the path starts with http:/https:// instead? It's certainly much cleaner

NEDJIMAbelgacem replied:
That's a really good idea, I will change to using this. Thanks

NEDJIMAbelgacem committed:
remove the use of dataSourceType

NEDJIMAbelgacem committed:
address reviews

NEDJIMAbelgacem committed:
remove unnecessary changes

nyalldawson commented on src/core/pointcloud/qgspointcloudlayer.h:
```diff
@@ -185,6 +185,8 @@ class CORE_EXPORT QgsPointCloudLayer : public QgsMapLayer
 
     QgsPointCloudLayerElevationProperties *mElevationProperties = nullptr;
+
+    QString mDataSourceType;
```
This is not used anywhere, is it?

nyalldawson commented:
QgsEptProviderMetadata::encodeUri and QgsEptProviderMetadata::decodeUri will need updating to handle http(s) uris -- in this case the "path" key isn't appropriate, since the uri isn't a file path. @nirvn do we have a convention here of key name for remote http URLs?

nirvn commented:
Not only do we not have a convention, but ATM both OGR and GDAL will keep non file paths into the "path" element. So this PR matches what we do elsewhere.

We should address this eventually, but neve had the need for an implementation so far.

NEDJIMAbelgacem commented:
> Not only do we not have a convention, but ATM both OGR and GDAL will keep non file paths into the "path" element. So this PR matches what we do elsewhere.
> 
> We should address this eventually, but neve had the need for an implementation so far.


For more context in the implementation, I didn't introduce a new provider class for the remote EPT. Only the index is setup differently for each case (remote vs local file).

NEDJIMAbelgacem committed:
Address reviews

NEDJIMAbelgacem committed:
On demand hierarchy loading in remote EPT

NEDJIMAbelgacem committed:
fix temporary file path

NEDJIMAbelgacem committed:
make QgsRemoteEptPointCloudIndex derive from QgsEptPointCloudIndex

NEDJIMAbelgacem committed:
refactor schema and manifest loading

NEDJIMAbelgacem committed:
Try to compilation errors

PeterPetrik commented on src/gui/providers/qgspointcloudsourceselect.cpp:
```diff
@@ -35,27 +42,114 @@ QgsPointCloudSourceSelect::QgsPointCloudSourceSelect( QWidget *parent, Qt::Windo
 void QgsPointCloudSourceSelect::addButtonClicked()
 {
-  if ( mPath.isEmpty() )
+  qDebug() << __PRETTY_FUNCTION__ << mDataSourceType;
```
remove

PeterPetrik commented on src/gui/providers/qgspointcloudsourceselect.cpp:
```diff
@@ -35,27 +42,114 @@ QgsPointCloudSourceSelect::QgsPointCloudSourceSelect( QWidget *parent, Qt::Windo
-                              tr( "No layers selected." ) );
-    return;
-  }
+    qDebug() << __PRETTY_FUNCTION__ << " : " << mPath;
```
remove

PeterPetrik commented on src/gui/providers/qgspointcloudsourceselect.cpp:
```diff
@@ -35,27 +42,114 @@ QgsPointCloudSourceSelect::QgsPointCloudSourceSelect( QWidget *parent, Qt::Windo
+      // maybe we should raise an assert if preferredProviders size is 0 or >1? Play it safe for now...
+      if ( preferredProviders.empty() )
+        continue;
+      qDebug() << "preferredProviders.at( 0 ).metadata()->key(): " << preferredProviders.at( 0 ).metadata()->key();
```
remove

PeterPetrik reviewed (commented):
1. If I place URL without ept.json at the end and add "Add" button it does nothing, no error, no added layer, etc. I suggest to try to add "ept.json" at the end if not present, or at least show some error to the user (e.g. try `https://s3-us-west-2.amazonaws.com/usgs-lidar-public/AK_NomeCreek_2010/ept.json` vs `https://s3-us-west-2.amazonaws.com/usgs-lidar-public/AK_NomeCreek_2010`)

2. crash (load `AK_NomeCreek_2010/ept.json`), change to classification 2d renderer, zoom-out 💥 
```
src/core/qgsmessagelog.cpp:29 : (logMessage) [1583030ms] [thread:0x7ff738467120] 2021-03-29T10:32:16 Network[1] Error transferring https://s3-us-west-2.amazonaws.com/usgs-lidar-public/AK_NomeCreek_2010/ept-sources/manifest.json - server replied: Not Found
src/core/pointcloud/qgspointcloudlayerrenderer.cpp:285 : (renderNodesAsync) [4069ms] [thread:0x7ff74e9d1aa0] Downloaded in : 3469ms
src/core/pointcloud/qgspointcloudlayerrenderer.cpp:285 : (renderNodesAsync) [0ms] [thread:0x7ff7389586a0] Downloaded in : 3966ms
Critical: Caught unhandled std::exception:  No LASzip VLR was found in the VLRs section
src/core/pointcloud/qgspointcloudlayerrenderer.cpp:285 : (renderNodesAsync) [135615ms] [thread:0x7ff74e5edce0] Downloaded in : 1373ms
src/core/pointcloud/qgspointcloudlayerrenderer.cpp:285 : (renderNodesAsync) [760ms] [thread:0x7ff72f4318d0] Downloaded in : 556ms
src/core/pointcloud/qgspointcloudlayerrenderer.cpp:285 : (renderNodesAsync) [2879ms] [thread:0x7ff7ae70f8c0] Downloaded in : 560ms
src/core/pointcloud/qgspointcloudlayerrenderer.cpp:285 : (renderNodesAsync) [1847ms] [thread:0x7ff738932520] Downloaded in : 591ms
Critical: Caught unhandled std::exception:  No LASzip VLR was found in the VLRs section
Critical: Caught unhandled std::exception:  The chunk table version number is unknown
Assertion failed: (sym < (1<<16)), function readShort, file /Users/peter/Projects/pointclouds/QGIS/external/laz-perf/decoder.hpp, line 236.
Abort trap: 6
```

wonder-sk commented on src/core/pointcloud/qgspointcloudblockhandle.h:
```diff
@@ -0,0 +1,59 @@
+    QgsPointCloudBlockHandle( const QString &dataType, const QgsPointCloudAttributeCollection &attributes, const QgsPointCloudAttributeCollection &requestedAttributes, QgsTileDownloadManagerReply *tileDownloadManagerReply );
+  signals:
+    //! Emitted when the block is loaded successfully
+    void blockLoadingSucceeded( QgsPointCloudBlock *block );
```
Passing a pointer to a new instance that should be deleted by caller is not a great pattern with signals/slots: in theory there could be multiple receivers connected to the signal and then it's unclear who should delete the pointer. A better solution is to make the block owned by QgsPointCloudBlockHandle class as a member variable (ideally unique_ptr).

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerrenderer.cpp:
```diff
@@ -203,15 +229,99 @@ bool QgsPointCloudLayerRenderer::render()
+  for ( int i = 0; i < nodes.size(); ++i )
+  {
+    if ( blocks[ i ] )
+      delete blocks[ i ];
```
As written in a comment above, let's make block owned by QgsPointCloudBlockHandle class, so that we do not need to worry about deleting block explicitly here.

wonder-sk commented on src/core/pointcloud/qgspointcloudblockhandle.h:
```diff
@@ -0,0 +1,59 @@
+    QString mDataType;
+    QgsPointCloudAttributeCollection mAttributes;
+    QgsPointCloudAttributeCollection mRequestedAttributes;
+    QgsTileDownloadManagerReply *mTileDownloadManagetReply = nullptr;
```
```suggestion
    QgsTileDownloadManagerReply *mTileDownloadManagerReply = nullptr;
```

wonder-sk commented on src/core/pointcloud/qgspointcloudblockhandle.cpp:
```diff
@@ -0,0 +1,63 @@
+  : mDataType( dataType ), mAttributes( attributes ), mRequestedAttributes( requestedAttributes ), mTileDownloadManagetReply( tileDownloadManagerReply )
+{
+  connect( mTileDownloadManagetReply, &QgsTileDownloadManagerReply::finished, this, &QgsPointCloudBlockHandle::blockFinishedLoading );
+}
```
hmm it does not seem that the mTileDownloadManagetReply object would get deleted...

NEDJIMAbelgacem replied:
That's true, I fixed that.

wonder-sk reviewed (commented):
Great stuff :tada:

A couple of comments though we should address before merging.

When testing I was often getting rendering stuck... but let's address the code comments and maybe that will also make the potential issues where the code could get stuck more apparent...

NEDJIMAbelgacem committed:
remove qDebug

NEDJIMAbelgacem commented on src/core/pointcloud/qgsremoteeptpointcloudindex.cpp:
```diff
@@ -0,0 +1,291 @@
+    return nullptr;  // unsupported
+  }
+
+  QgsFileDownloader downloader( fileUrl, fileName );
```
This is because if I use QgsTileDownloadManager I need to redirect the downloaded data to a file, so here I just download the file directly from fileUrl to fileName path (skips the additional explicit writing to the file).

NEDJIMAbelgacem committed:
Address Martin reviews

NEDJIMAbelgacem commented on src/core/pointcloud/qgsremoteeptpointcloudindex.cpp:
```diff
@@ -0,0 +1,291 @@
+    if ( mHierarchy.contains( n2 ) )
+      lst.append( n2 );
+  }
+  return lst;
```
I forgot that I can do that in C++ actually :( Thanks for the remainder

NEDJIMAbelgacem committed:
fix layout

NEDJIMAbelgacem committed:
add /ept.json is missing message box and refactor nodeData

NEDJIMAbelgacem committed:
make the loading then rendering steps done on a group basis

NEDJIMAbelgacem committed:
fix bug with nodes count is small

NEDJIMAbelgacem committed:
Fix node children and hasNode bug
Add caching for network requests
Fix requesting a file that is clearly not in the server

NEDJIMAbelgacem committed:
Fix the use of the same file name in decompressLaz
Fix block ownership issue

NEDJIMAbelgacem committed:
revert qgsfiledownloader changes

NEDJIMAbelgacem committed:
Add QgsFeedback to handle canceling rendering

NEDJIMAbelgacem committed:
fix tests

wonder-sk commented on src/core/pointcloud/qgseptpointcloudindex.h:
```diff
@@ -47,25 +47,28 @@ class CORE_EXPORT QgsEptPointCloudIndex: public QgsPointCloudIndex
     QString mDirectory;
     QString mWkt;
-
     int mPointCount = 0;
```
Actually we need to change from `int` to `qint64` ... `int` can't store more than ~2 billion points - not enough :-)

wonder-sk commented on src/core/pointcloud/qgseptdecoder.cpp:
```diff
@@ -466,4 +478,24 @@ QgsPointCloudBlock *QgsEptDecoder::decompressLaz( const QString &filename,
+    const QgsPointCloudAttributeCollection &attributes,
+    const QgsPointCloudAttributeCollection &requestedAttributes )
+{
+  QString filename = QDir::tempPath() + QDir::separator() + QStringLiteral( "%1.laz" ).arg( nodeStr );
```
naming files like this is not unique enough... consider 1. there could be multiple point cloud layers loaded at once, 2. there could be multiple QGIS instances running.

A good idea may be to use QTemporaryFile which guarantees safe naming of temporary files - and has functionality to auto-remove the temp files.

wonder-sk commented on src/core/pointcloud/qgseptdecoder.cpp:
```diff
@@ -466,4 +478,24 @@ QgsPointCloudBlock *QgsEptDecoder::decompressLaz( const QString &filename,
+    return nullptr;
+  }
+  return QgsEptDecoder::decompressLaz( filename, attributes, requestedAttributes );
+}
```
We are not removing the temporary files after they are not needed anymore... my temp folder is filling up with .laz files :smile:

wonder-sk commented on src/core/pointcloud/qgseptdecoder.cpp:
```diff
@@ -466,4 +478,24 @@ QgsPointCloudBlock *QgsEptDecoder::decompressLaz( const QString &filename,
+    const QgsPointCloudAttributeCollection &requestedAttributes )
+{
+  QString filename = QDir::tempPath() + QDir::separator() + QStringLiteral( "%1.laz" ).arg( nodeStr );
+  std::ofstream file( filename.toStdString(), std::ios::binary | std::ios::out );
```
No luck with string stream? If laz-perf does not like that, can you please prepare a test code to demonstrate the issue (e.g. on pastebin / github gist)

NEDJIMAbelgacem replied:
I was trying a different class previously but now using std::istringstream did work.

wonder-sk commented on src/core/pointcloud/qgseptpointcloudindex.cpp:
```diff
@@ -81,7 +126,7 @@ bool QgsEptPointCloudIndex::loadSchema( QFile &f )
 
   mSpan = result.value( QLatin1String( "span" ) ).toInt();
-  mPointCount = result.value( QLatin1String( "points" ) ).toInt();
+  mPointCount = std::min( ( double )std::numeric_limits<int>::max(), result.value( QLatin1String( "points" ) ).toDouble() );
```
Once we move to qint64, this min/max stuff should not be needed...

wonder-sk commented on src/core/pointcloud/qgspointcloudblockrequest.h:
```diff
@@ -0,0 +1,75 @@
+
+    /**
+     * Returns the requested block. if the returned block is nullptr, that means the data request failed
+     * Note: It is the responsablitiy of the caller to delete the block if it was loaded correctly
```
Why switching back to giving the responsibility of the caller to delete the block? :thinking:

NEDJIMAbelgacem replied:
In QgsRemoteEptPointCloudIndex::nodeData we internally use QgsPointCloudBlockRequest and take QgsPointCloudBlock from it and return the block to the caller, so depending on the situation we don't want QgsPointCloudBlock to be owned by QgsPointCloudBlockRequest.

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerrenderer.cpp:
```diff
@@ -203,15 +229,92 @@ bool QgsPointCloudLayerRenderer::render()
 
-  mReadyToCompose = true;
-  return !canceled;
+  const int groupSize = 4;
```
I am not sure if I understand why we are doing groups here - it would be worth adding some comments to explain that... and is that even needed? It complicates the code...

NEDJIMAbelgacem replied:
The previous initial implementation was to load all blocks in parallel and then render them sequentially. So I adjusted the implementation to load groupSize blocks in parallel and then render these groupSize blocks sequentially. It is needed since we might end up loading too many blocks which will make QGIS look non responsive when it's loading the blocks.
I will add an explanation, Do you think 4 is a good groupSize or should I change it to 8 or 16 ?

wonder-sk reviewed (commented):
Lazy loading of hierarchy does not seem to work correctly - in my tests with https://s3-us-west-2.amazonaws.com/usgs-lidar-public/NY_NewYorkCity/ept.json this only fetches 0-0-0-0.json but it never fetches anything deeper (for nodes containing -1 as the point count)

The GUI also seems to be confused when trying to use the above mentioned EPT dataset with ~5B points - due to integer overflow - I am unable to adjust point budget:
![image](https://user-images.githubusercontent.com/193367/113515426-c2b02200-9574-11eb-9ec5-bf6a0bc60b60.png)

NEDJIMAbelgacem committed:
change the point budge to qint64 and double spin box

NEDJIMAbelgacem committed:
switch to use QTemporaryFile

NEDJIMAbelgacem committed:
Add goupSize explanation

NEDJIMAbelgacem committed:
Use istringstream instead of creating a temp file

NEDJIMAbelgacem committed:
Fix hierarchy loading

NEDJIMAbelgacem committed:
classify the QgsPointCloudAttributeByRampRenderer color ramp shader by default

NEDJIMAbelgacem committed:
set more appropriate layer name

NEDJIMAbelgacem committed:
fix spelling

wonder-sk reviewed (approved):
Things are looking good to me now... Let's merge it!
